gpu/octree/src/cuda/approx_nsearch.cu

```diff
@@ -105,176 +113,96 @@ namespace pcl { namespace device { namespace appnearest_search
-				unsigned int laneId = Warp::laneId();
-				unsigned int warpId = threadIdx.x/warpSize;            
-
-				int active_lane = __ffs(mask) - 1; //[0..31]                        
-				mask &= ~(1 << active_lane);   
+			{
+				const unsigned int laneId = Warp::laneId();
 
-				volatile int* warp_buffer = &per_warp_buffer[warpId];
+				const int active_lane = __ffs(mask) - 1; //[0..31]
+				mask &= ~(1 << active_lane);
 
-				//broadcast beg
+				//broadcast beg and end
+				int fbeg, fend;
 				if (active_lane == laneId)
-					*warp_buffer = batch.octree.begs[node_idx];                    
-				int beg = *warp_buffer;
-
-				//broadcast end
-				if (active_lane == laneId)
-					*warp_buffer = batch.octree.ends[node_idx];
-				int end = *warp_buffer;
-
-				float3 active_query;
-				volatile float* warp_buffer_float = (float*)&per_warp_buffer[warpId];
+				{
+					fbeg = batch.octree.begs[node_idx];
+					fend = batch.octree.ends[node_idx];
+				}
+				const int beg = __shfl_sync(0xFFFFFFFF, fbeg, active_lane);
+				const int end = __shfl_sync(0xFFFFFFFF, fend, active_lane);
 
 				//broadcast warp_query
-				if (active_lane == laneId)
-					*warp_buffer_float = query.x;
-				active_query.x = *warp_buffer_float;
+				const float3 active_query = make_float3(
+					__shfl_sync(0xFFFFFFFF, query.x, active_lane),
+					__shfl_sync(0xFFFFFFFF, query.y, active_lane),
+					__shfl_sync(0xFFFFFFFF, query.z, active_lane)
+				);
 
-				if (active_lane == laneId)
-					*warp_buffer_float = query.y;
-				active_query.y = *warp_buffer_float;
+				const auto nearestPoint = NearestWarpKernel<KernelPolicy::CTA_SIZE>(beg, batch.points_step, end - beg, active_query);
 
 				if (active_lane == laneId)
-					*warp_buffer_float = query.z;
-				active_query.z = *warp_buffer_float;                            
-
-				int offset = NearestWarpKernel<KernelPolicy::CTA_SIZE>(batch.points + beg, batch.points_step, end - beg, active_query);                    
-
-				if (active_lane == laneId)
-					result_idx = beg + offset;
+					result_idx = beg + nearestPoint.first;
 			}
 		}
 
-        template<int CTA_SIZE>
-		__device__ __forceinline__ int NearestWarpKernel(const float* points, int points_step, int length, const float3& active_query)
-		{                        						
-            __shared__ volatile float dist2[CTA_SIZE];
-            __shared__ volatile int   index[CTA_SIZE];
-			
-            int tid = threadIdx.x;
-			dist2[tid] = std::numeric_limits<float>::max();
-
-			//serial step
-            for (int idx = Warp::laneId(); idx < length; idx += Warp::STRIDE)
-			{
-				float dx = points[idx                  ] - active_query.x;
-				float dy = points[idx + points_step    ] - active_query.y;
-				float dz = points[idx + points_step * 2] - active_query.z;
+		template <int CTA_SIZE>
+		__device__  std::pair<int, float>
+		NearestWarpKernel(const int beg,
+							const int field_step,
+							const int length,
+							const float3& active_query)
```

Comment:
Is the indentation here supposed to be like this? It looks okay in my IDE but weird here?

Reply:
Not really. This section still has tabs. How about passing the whole file through clang? You touched most of the code anyway.

Reply:
Oops, it isn't, fix incoming.

Reply:
oh okay, disregard the latest commit, didn't see your comment until now.